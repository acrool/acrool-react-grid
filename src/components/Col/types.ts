export enum EColType {auto='auto'}
export type TColumn = number | true | 'auto' | undefined;
