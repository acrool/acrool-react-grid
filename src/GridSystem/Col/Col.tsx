import React from 'react';
import styled, {css} from 'styled-components';
import {themeName} from '../../config';

import {TStyledProps, IColProps} from '../../types';

import {generateRWDStyled, cssGetter} from './utils';



/**
 * 判斷是否為空
 * @param value
 */
const suffix = (value: any) => {
    if (!!Number(value)) {
        return `-${value}`;
    } if (value === 'auto') {
        return '-auto';
    }
    return '';
};



/**
 * 產生 Debug 資訊
 * @param props
 */
const generateDebugData = (props: TStyledProps<IColProps>) => {
    return [
        props.col && `col${suffix(props.col)}`,
        props.sm && `col-sm${suffix(props.sm)}`,
        props.md && `col-md${suffix(props.md)}`,
        props.lg && `col-lg${suffix(props.lg)}`,
        props.xl && `col-xl${suffix(props.xl)}`,
        props.xxl && `col-xxl${suffix(props.xxl)}`,
    ]
        .filter(Boolean)
        .join(' ');
};




/**
 * Col Component
 * breakpoints sort: sm => xl
 *
 * qa: min-height=1px
 * https://css-tricks.com/make-sure-columns-dont-collapse-horizontally/
 *
 */
const Col = styled.div.attrs((props: TStyledProps<IColProps>) => ({
    'data-grid': generateDebugData(props),
}))`
  //box-sizing: border-box;
  //min-height: 1px;
  //min-width: 0; // 解決下層有使用 white-space: nowrap; 產生衝突跑版

  ${(props: TStyledProps<IColProps>) => css`
     ${props.col && cssGetter.col(props.col, props.theme[themeName]?.gridColumns)};
     ${generateRWDStyled(props)};
 `}
`;

export default Col;
