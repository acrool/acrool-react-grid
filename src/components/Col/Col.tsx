import React, {MouseEvent} from 'react';
import styled, {css} from 'styled-components/macro';

import getCss from './css';
import {TColumn} from './types';
import media from '../../media';
import {isEmpty, suffix} from '../../utils';
import {mediaSizes} from '../../config';
import {FCChildrenProps, TStyledProps} from '../../typings';


interface IProps extends FCChildrenProps{
    col?: TColumn;
    sm?: TColumn;
    md?: TColumn;
    lg?: TColumn;
    xl?: TColumn;
    xxl?: TColumn;
}



/**
 * 產生 Debug 資訊
 * @param props
 */
const generateDebugData = (props: TStyledProps<IProps>) => {
    if(process.env.NODE_ENV === 'production'){
        return undefined;
    }
    return [
        props.col && `col${suffix(props.col)}`,
        props.sm && `col-sm${suffix(props.sm)}`,
        props.md && `col-md${suffix(props.md)}`,
        props.lg && `col-lg${suffix(props.lg)}`,
        props.xl && `col-xl${suffix(props.xl)}`,
        props.xxl && `col-xxl${suffix(props.xxl)}`,
    ]
        .filter(Boolean)
        .join(' ');
};



/**
 * 產生 RWD 樣式
 * @param props
 */
const generateRWDStyled = (props: TStyledProps<IProps>) => {
    return mediaSizes
        .map(sizeName => {
            return media[sizeName]`
             padding-right: ${props.theme.styledGrid.gridGutterWidthMedia[sizeName]}px;
             padding-left: ${props.theme.styledGrid.gridGutterWidthMedia[sizeName]}px;

            ${!isEmpty(props[sizeName]) && css`
                ${getCss.col(props[sizeName], props.theme.styledGrid.gridColumns)};
            `}
    `;
        });
};

/**
 * Col Component
 * breakpoints sort: sm => xl
 *
 * qa: min-height=1px
 * https://css-tricks.com/make-sure-columns-dont-collapse-horizontally/
 *
 */
const Col = styled.div.attrs((props: TStyledProps<IProps>) => ({
    'data-grid': 'col',
    'data-debug': generateDebugData(props),
}))`
  box-sizing: border-box;
  position: relative;
  width: 100%;
  min-height: 1px;

  ${(props: TStyledProps<IProps>) => css`
     padding-right: ${props.theme.styledGrid.gridGutterWidth}px;
     padding-left: ${props.theme.styledGrid.gridGutterWidth}px;

     >[data-grid=row]{
        flex: 0 1 100%;
    }

     min-width: 0; // 解決下層有使用 white-space: nowrap; 產生衝突跑版
     ${props.col && getCss.col(props.col, props.theme.styledGrid.gridColumns)};

     ${generateRWDStyled(props)};
 `}
`;

export default Col;
