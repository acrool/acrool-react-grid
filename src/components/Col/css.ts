import {EColType, TColumn} from './types';

type ColCss = {
    col: (column: TColumn, gridColumns: number) => string
};

const css: ColCss = {
    col: (column, gridColumns) => {
        let colFlexBasis = 0;

        switch (column) {
        case true:
            return `
          display: block;
          flex-basis: 0;
          flex-grow: 1;
        `;
        case EColType.auto:
            return `
          display: block;
          flex: 0 0 auto;
          width: auto;
          max-width: none;
        `;
        case EColType.hidden:
            return `
          display: none;
        `;
        case EColType.visible:
            return `
          display: block;
        `;
        default:
            const columnNumber = (typeof column === 'number' ? column : 0);
            colFlexBasis = (100 / gridColumns) * columnNumber;
            return `
            flex: 0 0 ${colFlexBasis}%;
            width: ${colFlexBasis}%;
        `;
        }
    },
};

export default css;
