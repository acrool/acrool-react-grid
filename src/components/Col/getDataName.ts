import {suffix} from '../../utils';
import {ColProps} from './types';

export default (props: ColProps) => (process.env.NODE_ENV === 'production'
    ? undefined
    : [
        props.col && `col${suffix(props.col)}`,
        props.sm && `col-sm${suffix(props.sm)}`,
        props.md && `col-md${suffix(props.md)}`,
        props.lg && `col-lg${suffix(props.lg)}`,
        props.xl && `col-xl${suffix(props.xl)}`,
        props.xxl && `col-xxl${suffix(props.xxl)}`,
    ]
        .filter(Boolean)
        .join(' '));
