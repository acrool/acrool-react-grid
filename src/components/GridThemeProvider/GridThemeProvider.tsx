import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle} from 'styled-components';

import {defaultGridTheme, themeName} from '../../config';
import {IGridSetting} from '../../types';


interface IProps {
    gridTheme: Partial<IGridSetting>;
    children: React.ReactNode;
}

const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    
`;
