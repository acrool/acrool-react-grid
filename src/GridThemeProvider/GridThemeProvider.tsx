import React, {PropsWithChildren} from 'react';
import {ThemeProvider, createGlobalStyle, css} from 'styled-components';

import {defaultGridTheme, themeName} from '../config';
import {TStyledProps, TThemeProps, IGridThemeProviderProps} from '../types';
import {generateRWDStyled} from './utils';
import '../styles.css';



const ThemeProviderFixed = ThemeProvider as unknown as React.FC<PropsWithChildren<{ theme: any }>>;


/**
 * Grid Theme Provider
 * use :
 *     <GridThemeProvider gridTheme={{
 *         gridGutterWidth: 5,
 *     }}>
 *          <Container>, <Row>, <Col> ...then eat this setting
 *     </GridThemeProvider>
 */
const GridThemeProvider = (props: IGridThemeProviderProps) => {
    const {gridTheme, children} = props;
    const {gridBreakpoints, containerMaxWidths, gridGutterWidthMedia, ...baseDefaultTheme} = defaultGridTheme;

    // console.log('test render bear-react-grid version 10');
    const composeGridTheme = {
        [themeName]: {
            ...baseDefaultTheme,
            ...gridTheme,
            gridBreakpoints: Object.assign({}, gridBreakpoints, gridTheme?.gridBreakpoints),
            containerMaxWidths: Object.assign({}, containerMaxWidths, gridTheme?.containerMaxWidths),
            gridGutterWidthMedia: Object.assign({}, gridGutterWidthMedia, gridTheme?.gridGutterWidthMedia),
        }
    };

    return <ThemeProviderFixed theme={composeGridTheme}>
        {children}
        <GlobalStyled/>
    </ThemeProviderFixed>;
};

export default GridThemeProvider;


const GlobalStyled = createGlobalStyle`
    :root{
        --bear-spacer: 1rem;
        --bear-spacer-1: calc(1rem * .25);
        --bear-spacer-2: calc(1rem * .5);
        --bear-spacer-3: calc(1rem);
        --bear-spacer-4: calc(1rem * 1.5);
        --bear-spacer-5: calc(1rem * 3);
    }

    ${(props: TStyledProps<TThemeProps>) => css`
        ${generateRWDStyled(props)}
    `}

    [data-row][data-horizontal="left"]{
        justify-content: flex-start;
    }
    [data-row][data-horizontal="center"]{
        justify-content: center;
    }
    [data-row][data-horizontal="right"]{
        justify-content: flex-end;
    }
    [data-row][data-vertical="top"]{
        justify-content: flex-start;
    }
    [data-row][data-vertical="center"]{
        justify-content: center;
    }
    [data-row][data-vertical="bottom"]{
        justify-content: flex-end;
    }
`;
