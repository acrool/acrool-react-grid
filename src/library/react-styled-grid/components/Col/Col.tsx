import React from 'react';
import styled, {css} from 'styled-components/macro';

import getCss from './css';
import getDataName from './getDataName';
import media from '../../media';
import {isEmpty} from '../../utils';
import {themeName} from '../../config';
import {ColProps} from './types';
import {themeProps} from '../ThemeProvider/types';

type Props = ColProps & {
    theme: themeProps;
};

// eslint-disable-next-line array-callback-return
const generateMedia = (props: any) => Object
    .keys(props.theme.reactStyledGrid.gridBreakpoints)
    .filter(sizeName => sizeName !== 'xs')
    .map(sizeName => {
        return media[sizeName]`
             padding-right: ${props.theme.reactStyledGrid.gridGutterWidthMedia[sizeName]}px;
             padding-left: ${props.theme.reactStyledGrid.gridGutterWidthMedia[sizeName]}px;

            ${!isEmpty(props[sizeName]) && css`
                ${getCss.col(props[sizeName], props.theme.reactStyledGrid.gridColumns)};
            `}
    `;
    });

/**
 * Col Component
 * breakpoints sort: sm => xl
 *
 * qa: min-height=1px
 * https://css-tricks.com/make-sure-columns-dont-collapse-horizontally/
 *
 */
const Col: any = styled.div.attrs((props: Props) => ({
    'data-grid': 'col',
    'data-debug': getDataName(props),
}))`
  position: relative;
  width: 100%;
  min-height: 1px;

  ${(props: any) => css`
     padding-right: ${props.theme.reactStyledGrid.gridGutterWidth}px;
     padding-left: ${props.theme.reactStyledGrid.gridGutterWidth}px;

     >[data-grid=row]{
        flex: 0 1 100%;
    }

     min-width: 0; // 解決下層有使用 white-space: nowrap; 產生衝突跑版
     ${props.col && getCss.col(props.col, props.theme.reactStyledGrid.gridColumns)};

     ${generateMedia(props)};
 `}
`;

export default ({forwardRef, forwardAs, ...props}: ColProps) => <Col ref={forwardRef} as={forwardAs} {...props}/>;
