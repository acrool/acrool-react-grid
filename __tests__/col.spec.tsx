import React from 'react';
import 'jest-styled-components';
import {render} from '@testing-library/react';
import '@testing-library/dom';
import Col from '../src/GridSystem/Col';
import GridThemeProvider from '../src/GridThemeProvider';



describe('Col', () => {
    it('renders col-true', () => {
        const {container} = render(<Col col>col</Col>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders col-auto', () => {
        const {container} = render(<Col col="auto">col</Col>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });

    it('renders col-6', () => {
        const {container} = render(<Col col={6}>col</Col>, {wrapper: GridThemeProvider});
        expect(container.firstChild).toMatchSnapshot();
    });
});
