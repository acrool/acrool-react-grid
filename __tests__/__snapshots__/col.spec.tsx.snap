// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Col renders col-6 1`] = `
.c0 {
  min-height: 1px;
  min-width: 0;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: 50%;
}

<div
  class="c0"
  data-grid="col-6"
>
  col
</div>
`;

exports[`Col renders col-auto 1`] = `
.c0 {
  min-height: 1px;
  min-width: 0;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  width: auto;
}

<div
  class="c0"
  data-grid="col-auto"
>
  col
</div>
`;

exports[`Col renders col-true 1`] = `
.c0 {
  min-height: 1px;
  min-width: 0;
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
}

<div
  class="c0"
  data-grid="col-true"
>
  col
</div>
`;
